In [2]:
import pandas as pd

## Data Import 
First, we'll import the data and break each paragraph into a row of our dataframe.

In [3]:
with open('C:/Users/kruss/Downloads/Copy of Book 1 (text data) (1).txt', 'r', encoding='utf-8') as file:
    data = file.read()

rows = data.split('\n')

df = pd.DataFrame(rows, columns=['text'])

print(df)

                                                   text
0                                               PART 1
1     Quinn buried her head in her hands and looked ...
2             To her credit, Quinn had tried her har...
3             She looked at the students around her....
4             It was a record-setting day for the zo...
...                                                 ...
3280  But if anyone can manage to stop, I think it’s...
3281  No matter what, you’re my superhero. The earli...
3282  So I guess it’s alright that things worked out...
3283                                             -Maeve
3284                                            – FIN –

[3285 rows x 1 columns]


In [4]:
df["book_num"] = 1
print(df)

                                                   text  book_num
0                                               PART 1         1
1     Quinn buried her head in her hands and looked ...         1
2             To her credit, Quinn had tried her har...         1
3             She looked at the students around her....         1
4             It was a record-setting day for the zo...         1
...                                                 ...       ...
3280  But if anyone can manage to stop, I think it’s...         1
3281  No matter what, you’re my superhero. The earli...         1
3282  So I guess it’s alright that things worked out...         1
3283                                             -Maeve         1
3284                                            – FIN –         1

[3285 rows x 2 columns]


In [5]:
chapter = 1
chp_list = list()

for i in range (0, len(df)):
    if df["text"].loc[i] == "./.":
        chapter = chapter + 1
    chp_list.append(chapter)

df["chapter"] = chp_list

In [6]:
paragraph = 1
pgh_list = list()
pgh_list.append(1)

for i in range (1, len(df)):
    if df["chapter"].loc[i] != df["chapter"].loc[i - 1]:
        paragraph = 1
    else:
        paragraph = paragraph + 1
    pgh_list.append(paragraph)

df["paragraph"] = pgh_list

In [7]:
print(df)

                                                   text  book_num  chapter  \
0                                               PART 1         1        1   
1     Quinn buried her head in her hands and looked ...         1        1   
2             To her credit, Quinn had tried her har...         1        1   
3             She looked at the students around her....         1        1   
4             It was a record-setting day for the zo...         1        1   
...                                                 ...       ...      ...   
3280  But if anyone can manage to stop, I think it’s...         1       46   
3281  No matter what, you’re my superhero. The earli...         1       46   
3282  So I guess it’s alright that things worked out...         1       46   
3283                                             -Maeve         1       46   
3284                                            – FIN –         1       46   

      paragraph  
0             1  
1             2  
2        

In [8]:
with open('C:/Users/kruss/Downloads/Copy of Book Two (text data) (2).txt', 'r', encoding='utf-8') as file:
    data = file.read()

rows = data.split('\n')

df2 = pd.DataFrame(rows, columns=['text'])

df2["book_num"] = 2

chapter = 1
chp_list = list()

for i in range (0, len(df2)):
    if df2["text"].loc[i] == "./.":
        chapter = chapter + 1
    chp_list.append(chapter)

df2["chapter"] = chp_list

paragraph = 1
pgh_list = list()
pgh_list.append(1)

for i in range (1, len(df2)):
    if df2["chapter"].loc[i] != df2["chapter"].loc[i - 1]:
        paragraph = 1
    else:
        paragraph = paragraph + 1
    pgh_list.append(paragraph)

df2["paragraph"] = pgh_list

print(df2)

                                                   text  book_num  chapter  \
0     The girl in the dark black dress with the dar...         2        1   
1             Next in line was her great-uncle, a lo...         2        1   
2     She hated him. But more crucially, she found h...         2        1   
3             He hugged her mother and lamented into...         2        1   
4                   But she liked it better, this way.          2        1   
...                                                 ...       ...      ...   
2339  She slid her fingers around, looking for the s...         2       26   
2340  She began to feel sick. She reached for the si...         2       26   
2341                                 Nothing, nothing.          2       26   
2342  She pressed down, touched her jaw, went back t...         2       26   
2343           A realm between the living and the dead.         2       26   

      paragraph  
0             1  
1             2  
2        

In [9]:
df = pd.concat([df, df2], ignore_index=True)

In [10]:
df

text  book_num  chapter  \
0                                               PART 1         1        1   
1     Quinn buried her head in her hands and looked ...         1        1   
2             To her credit, Quinn had tried her har...         1        1   
3             She looked at the students around her....         1        1   
4             It was a record-setting day for the zo...         1        1   
...                                                 ...       ...      ...   
5624  She slid her fingers around, looking for the s...         2       26   
5625  She began to feel sick. She reached for the si...         2       26   
5626                                 Nothing, nothing.          2       26   
5627  She pressed down, touched her jaw, went back t...         2       26   
5628           A realm between the living and the dead.         2       26   

      paragraph  
0             1  
1             2  
2             3  
3             4  
4             5  
...         ...  
5624         24  
5625         25  
5626         26  
5627         27  
5628         28  

[5629 rows x 4 columns]

## Preprocessing
Now that we have our corpus into a workable format, we can preprocess the text data. 

In [55]:
import pandas as pd
import numpy as np

from gensim import models
from gensim import corpora 
from gensim.corpora import Dictionary
from gensim.models import LdaModel
from gensim.models.coherencemodel import CoherenceModel

import nltk

STOPLIST = list(nltk.corpus.stopwords.words('english'))

In [56]:
print(len(df))
np.random.seed(2)
df.sample(10)

2592


text  book_num  chapter  \
1810          It was easy, really. Walk forward, tur...         1       30   
2556  Maeve was huddled on the ground, shaking and w...         1       40   
4945          “Okay. So, uh, where to start. It’s Su...         2       18   
153   Quinn found her trust in her parents fading qu...         1        3   
2007          “Hmmm…” said Maeve, desperately trying...         1       34   
682   She started to sweat. She had mere seconds lef...         1        9   
114           “Dude, I’m just messing with you. I su...         1        2   
269           Quinn had to breathe. She threw open t...         1        4   
4735  Just as she opened her mouth to say something ...         2       16   
2371  “Come, now,” said President Dorothy, helping M...         1       38   

      paragraph                                   tokens_stopwords  \
1810          6  [it, was, easy, really, walk, forward, turn, t...   
2556         43  [maeve, was, huddled, on, the, ground, shaking...   
4945        119  [okay, so, uh, where, to, start, it, s, sunday...   
153           4  [quinn, found, her, trust, in, her, parents, f...   
2007         62  [said, maeve, desperately, trying, to, decide,...   
682          67  [she, started, to, sweat, she, had, mere, seco...   
114          28  [dude, i, m, just, messing, with, you, i, suck...   
269          78  [quinn, had, to, breathe, she, threw, open, th...   
4735         50  [just, as, she, opened, her, mouth, to, say, s...   
2371         88  [come, now, said, president, dorothy, helping,...   

                                    tokens_no_stopwords  
1810  [easy, really, walk, forward, turn, handle, si...  
2556  [maeve, huddled, ground, shaking, whimpering, ...  
4945  [okay, uh, start, sunday, night, actually, ear...  
153   [quinn, found, trust, parents, fading, quickly...  
2007  [said, maeve, desperately, trying, decide, nam...  
682   [started, sweat, mere, seconds, left, going, b...  
114   [dude, messing, suck, calculus, even, know, in...  
269   [quinn, breathe, threw, open, door, stepped, o...  
4735  [opened, mouth, say, something, nicholas, aman...  
2371  [come, said, president, dorothy, helping, maev...

There are some very short paragraphs here. Sometimes, a paragraph will be just one sentence or a few words long. We will delete all potential documents that have less than 100 characters. 

In [57]:
df_original = df
df = df[df['text'].str.len() >= 100]

In [58]:
print(len(df))
df.sample(10)

2592


text  book_num  chapter  \
2965          “You won’t be here forever. But all I ...         1       44   
3023          Her body, limp, unresponsive, was thro...         1       45   
3382  Her heart sank, and Dan must have noticed, bec...         2        3   
1188  He sat against the rack of cleaning supplies, ...         1       16   
4947          “As you might imagine, the last two da...         2       18   
4765          This is it, this is the end, this is i...         2       17   
3494          There were five of them, all together,...         2        5   
1882          She opened her mouth to respond, but t...         1       32   
3038  The eye of the storm of panic she felt was a p...         1       45   
336   “And if that big-ass knife is for me, then I g...         1        5   

      paragraph                                   tokens_stopwords  \
2965        133  [you, won, t, be, here, forever, but, all, i, ...   
3023         14  [her, body, limp, unresponsive, was, thrown, o...   
3382         39  [her, heart, sank, and, dan, must, have, notic...   
1188          4  [he, sat, against, the, rack, of, cleaning, su...   
4947        121  [as, you, might, imagine, the, last, two, days...   
4765         17  [this, is, it, this, is, the, end, this, is, i...   
3494         29  [there, were, five, of, them, all, together, a...   
1882         34  [she, opened, her, mouth, to, respond, but, th...   
3038         29  [the, eye, of, the, storm, of, panic, she, fel...   
336          27  [and, if, that, knife, is, for, me, then, i, g...   

                                    tokens_no_stopwords  
2965  [forever, tell, talk, others, way, talking, ey...  
3023  [body, limp, unresponsive, thrown, onto, makes...  
3382  [heart, sank, dan, must, noticed, practically,...  
1188  [sat, rack, cleaning, supplies, hands, grippin...  
4947  [might, imagine, last, two, days, pretty, inte...  
4765   [end, alone, like, always, years, plus, forever]  
3494  [five, together, shaking, hand, finally, stumb...  
1882  [opened, mouth, respond, sound, late, bell, lo...  
3038  [eye, storm, panic, felt, poisonously, allurin...  
336   [knife, guess, made, easier, target, showing, ...

Now, we have 2,592 documents that comprise our corpus. From here, we can clean and process our text. 

In [59]:
def process_text_nltk(
    text: str,
    lower_case: bool = True,
    remove_punct: bool = True,
    remove_stopwords: bool = False,
    lemma: bool = False,
    string_or_list: str = "str",
):
    tokens = nltk.word_tokenize(text)

    if lower_case:
        tokens = [token.lower() if token.isalpha() else token for token in tokens]

    if remove_punct:
        tokens = [token for token in tokens if token.isalpha()]

    if remove_stopwords:
        tokens = [token for token in tokens if not token in STOPLIST]

    if lemma:
        tokens = [nltk.wordnet.WordNetLemmatizer().lemmatize(token) for token in tokens]

    if string_or_list != "list":
        doc = " ".join(tokens)
    else:
        doc = tokens

    return doc

In [60]:
df["tokens_stopwords"] = df.text.apply(process_text_nltk, remove_stopwords = False, string_or_list = "list")
df["tokens_no_stopwords"] = df.text.apply(process_text_nltk, remove_stopwords = True, string_or_list = "list")

In [62]:
docs_list_stopwords = list(df.tokens_stopwords)
docs_list_no_stopwords = list(df.tokens_no_stopwords)

In [61]:
from gensim.models import phrases

In [63]:


bigram_model = Phrases(docs_list_stopwords, min_count=1, delimiter = "_", connector_words = phrases.ENGLISH_CONNECTOR_WORDS)

# Print bigram in document
for token in bigram_model[docs_list_stopwords[10]]:
    if "_" in token:
        print(token)

NameError: name 'Phrases' is not defined